File: paleo_workbench/ui/layout_persistence.py
"""QSettings-backed persistence for panel float/dock layouts."""

from __future__ import annotations

import logging
from dataclasses import dataclass

from PySide6.QtCore import QSettings, QRect

_log = logging.getLogger(__name__)

#: 统一的 QSettings 身份（B2）：历史上 shell 用 ``WorkstationV3``、面板浮
#: 动用 ``paleo-workbench``、主题用 ``Workstation`` 三处分裂；统一到主题
#: 已在用的 (PaleoWorkbench, Workstation)，旧数据一次性迁移。
SETTINGS_ORG = "PaleoWorkbench"
SETTINGS_APP = "Workstation"

#: 布局状态版本：不认识（更新或缺失）的版本一律丢弃走默认布局。
LAYOUT_STATE_VERSION = 4

_WINDOW_STATE_KEY = "layout/window_state"
_STATE_VERSION_KEY = "layout/state_version"
_INSPECTOR_FLAG_KEY = "layout/inspector_user_hidden"

#: 旧 shell 身份（WorkstationV3）下的窗口状态键。
_LEGACY_SHELL_APP = "WorkstationV3"
_LEGACY_WINDOW_STATE_KEY = "layout/windowState.v4"
#: 旧面板浮动持久化身份。
_LEGACY_PANELS_APP = "paleo-workbench"


def migrate_legacy_layout_settings() -> bool:
    """把历史 QSettings 身份里的布局数据迁到 (PaleoWorkbench, Workstation)。

    纯函数（只依赖默认 QSettings 存储），幂等：新键已有值时不覆盖，
    迁移成功后删除旧键。返回是否有任何数据被迁移——供测试与诊断断言。

    迁移内容：

    - ``(PaleoWorkbench, WorkstationV3)`` 的 ``layout/windowState.v4`` 与
      ``layout/inspector_user_hidden`` → 新应用名下 ``layout/window_state``
      /同名键，并补写 ``layout/state_version = 4``；
    - ``(PaleoWorkbench, paleo-workbench)`` 的 ``panel_layout`` 组 → 新应用
      名下同名组（键前缀不变）。
    """
    migrated = False
    target = QSettings(SETTINGS_ORG, SETTINGS_APP)

    legacy_shell = QSettings(SETTINGS_ORG, _LEGACY_SHELL_APP)
    legacy_shell.sync()
    state = legacy_shell.value(_LEGACY_WINDOW_STATE_KEY)
    if state is not None:
        if target.value(_WINDOW_STATE_KEY) is None:
            target.setValue(_WINDOW_STATE_KEY, state)
        inspector_hidden = legacy_shell.value(_INSPECTOR_FLAG_KEY, None)
        if inspector_hidden is not None:
            target.setValue(_INSPECTOR_FLAG_KEY, inspector_hidden)
        target.setValue(_STATE_VERSION_KEY, LAYOUT_STATE_VERSION)
        legacy_shell.remove(_LEGACY_WINDOW_STATE_KEY)
        legacy_shell.remove(_INSPECTOR_FLAG_KEY)
        migrated = True

    legacy_panels = QSettings(SETTINGS_ORG, _LEGACY_PANELS_APP)
    legacy_panels.sync()
    legacy_panels.beginGroup("panel_layout")
    try:
        keys = list(legacy_panels.allKeys())
        for key in keys:
            # 组内读、组外写：新存储键前缀保持 panel_layout/… 不变。
            target.setValue(f"panel_layout/{key}", legacy_panels.value(key))
    finally:
        legacy_panels.endGroup()
    if keys:
        legacy_panels.remove("panel_layout")
        migrated = True

    if migrated:
        target.sync()
        _log.info("migrated legacy layout settings into (%s, %s)", SETTINGS_ORG, SETTINGS_APP)
    return migrated


@dataclass(frozen=True)
class PanelLayoutRecord:
    """Everything persisted about one panel, keyed ``page:panel``."""

    floating: bool = False
    geometry: QRect | None = None
    docked_sizes: tuple[int, ...] | None = None
    visible: bool = True

    @property
    def is_empty(self) -> bool:
        """True when the store holds nothing worth restoring."""
        return not self.floating and not self.docked_sizes and self.visible


class LayoutPersistence:
    """Persist per-panel layout state (float geometry, docked sizes, visibility).

    Keys follow the ``page:panel`` convention (``"mapping:layer_tree"``) so
    persisted entries stay namespaced per page. Entries are only written when
    a float/dock actually happens, so a session that never floated a panel
    leaves nothing behind and restoring is a safe no-op — under offscreen CI
    nothing reads or writes unless a caller asks for it.

    Pass an explicit :class:`QSettings` to bind a custom backend (tests use a
    temporary ini file); the default binds the workbench org/application store
    lazily on first use.
    """

    ORGANIZATION = SETTINGS_ORG
    APPLICATION = SETTINGS_APP
    GROUP = "panel_layout"

    def __init__(self, settings: QSettings | None = None) -> None:
        self._settings = settings

    # --- write ----------------------------------------------------------

    def save_float(self, key: str, geometry: QRect) -> None:
        """Record ``key`` as floating at ``geometry`` (and visible)."""
        settings = self._bind()
        settings.beginGroup(self._group(key))
        try:
            settings.setValue("floating", True)
            settings.setValue("visible", True)
            settings.setValue(
                "geometry",
                "{},{},{},{}".format(
                    geometry.x(), geometry.y(), geometry.width(), geometry.height()
                ),
            )
        finally:
            settings.endGroup()
        settings.sync()

    def save_dock(self, key: str, sizes) -> None:
        """Record ``key`` as docked again with the splitter ``sizes``."""
        settings = self._bind()
        settings.beginGroup(self._group(key))
        try:
            settings.setValue("floating", False)
            settings.setValue("visible", True)
            settings.setValue(
                "docked_sizes", self._encode_sizes(sizes)
            )
        finally:
            settings.endGroup()
        settings.sync()

    def save_docked_sizes(self, key: str, sizes) -> None:
        """Record only the splitter ``sizes`` (float keeps ``floating`` True).

        Written at float time so a crash mid-float still leaves the pre-float
        dock slot recoverable.
        """
        settings = self._bind()
        settings.setValue(
            f"{self._group(key)}/docked_sizes", self._encode_sizes(sizes)
        )
        settings.sync()

    def save_visibility(self, key: str, visible: bool) -> None:
        """Record only the visibility flag (a floating panel was closed)."""
        settings = self._bind()
        settings.setValue(f"{self._group(key)}/visible", bool(visible))
        settings.sync()

    def clear(self, key: str) -> None:
        settings = self._bind()
        settings.remove(self._group(key))
        settings.sync()

    # --- read -----------------------------------------------------------

    def load(self, key: str) -> PanelLayoutRecord:
        settings = self._bind()
        settings.beginGroup(self._group(key))
        try:
            floating = settings.value("floating", False, type=bool)
            visible = settings.value("visible", True, type=bool)
            geometry = self._parse_geometry(
                settings.value("geometry", "", type=str)
            )
            sizes = self._parse_sizes(
                settings.value("docked_sizes", "", type=str)
            )
        finally:
            settings.endGroup()
        return PanelLayoutRecord(
            floating=floating, geometry=geometry, docked_sizes=sizes, visible=visible
        )

    # --- internals ------------------------------------------------------

    def _bind(self) -> QSettings:
        if self._settings is None:
            # 默认后端绑定前先做一次旧身份迁移（幂等）：孤立使用
            # LayoutPersistence（无工作站壳）也能读到旧面板布局。
            migrate_legacy_layout_settings()
            self._settings = QSettings(self.ORGANIZATION, self.APPLICATION)
        return self._settings

    @classmethod
    def _group(cls, key: str) -> str:
        return f"{cls.GROUP}/{key}"

    @staticmethod
    def _encode_sizes(sizes) -> str:
        return ",".join(str(int(size)) for size in sizes)

    @staticmethod
    def _parse_geometry(raw: str) -> QRect | None:
        parts = [part for part in (raw or "").split(",") if part]
        if len(parts) != 4:
            return None
        try:
            x, y, width, height = (int(part) for part in parts)
        except ValueError:
            return None
        return QRect(x, y, width, height)

    @staticmethod
    def _parse_sizes(raw: str) -> tuple[int, ...] | None:
        parts = [part for part in (raw or "").split(",") if part]
        if not parts:
            return None
        try:
            return tuple(int(part) for part in parts)
        except ValueError:
            return None

File: libs/ui_shell/src/layout_persistence.cpp
#include "pwb/ui_shell/layout_persistence.hpp"

#include <QLoggingCategory>
#include <QStringList>

namespace pwb::ui_shell {

namespace {
Q_LOGGING_CATEGORY(lcLayout, "pwb.ui_shell.layout")

constexpr const char* kWindowStateKey = "layout/window_state";
constexpr const char* kStateVersionKey = "layout/state_version";
constexpr const char* kInspectorFlagKey = "layout/inspector_user_hidden";
// Legacy shell identity (WorkstationV3) window-state key.
constexpr const char* kLegacyShellApp = "WorkstationV3";
constexpr const char* kLegacyWindowStateKey = "layout/windowState.v4";
// Legacy panel-floating persistence identity.
constexpr const char* kLegacyPanelsApp = "paleo-workbench";

QString qstr(const std::string& s) {
    return QString::fromUtf8(s.data(), static_cast<qsizetype>(s.size()));
}
}  // namespace

bool migrate_legacy_layout_settings() {
    bool migrated = false;
    QSettings target(QString::fromLatin1(kSettingsOrg),
                     QString::fromLatin1(kSettingsApp));

    // (PaleoWorkbench, WorkstationV3) window state + inspector flag.
    QSettings legacy_shell(QString::fromLatin1(kSettingsOrg),
                           QString::fromLatin1(kLegacyShellApp));
    legacy_shell.sync();
    const QVariant state = legacy_shell.value(QLatin1String(kLegacyWindowStateKey));
    if (state.isValid()) {
        if (!target.contains(QLatin1String(kWindowStateKey))) {
            target.setValue(QLatin1String(kWindowStateKey), state);
        }
        const QVariant inspector_hidden =
            legacy_shell.value(QLatin1String(kInspectorFlagKey));
        if (inspector_hidden.isValid()) {
            target.setValue(QLatin1String(kInspectorFlagKey), inspector_hidden);
        }
        target.setValue(QLatin1String(kStateVersionKey), kLayoutStateVersion);
        legacy_shell.remove(QLatin1String(kLegacyWindowStateKey));
        legacy_shell.remove(QLatin1String(kInspectorFlagKey));
        migrated = true;
    }

    // (PaleoWorkbench, paleo-workbench) panel_layout group.
    QSettings legacy_panels(QString::fromLatin1(kSettingsOrg),
                            QString::fromLatin1(kLegacyPanelsApp));
    legacy_panels.sync();
    legacy_panels.beginGroup(QStringLiteral("panel_layout"));
    const QStringList keys = legacy_panels.allKeys();
    for (const QString& key : keys) {
        // Read inside the group, write outside: new keys keep the
        // panel_layout/… prefix unchanged.
        target.setValue(QStringLiteral("panel_layout/") + key,
                        legacy_panels.value(key));
    }
    legacy_panels.endGroup();
    if (!keys.isEmpty()) {
        legacy_panels.remove(QStringLiteral("panel_layout"));
        migrated = true;
    }

    if (migrated) {
        target.sync();
        qCInfo(lcLayout,
               "migrated legacy layout settings into (%s, %s)", kSettingsOrg,
               kSettingsApp);
    }
    return migrated;
}

LayoutPersistence::LayoutPersistence(QSettings* settings)
    : settings_(settings) {}

void LayoutPersistence::save_float(const std::string& key,
                                   const QRect& geometry) {
    QSettings& settings = bind();
    settings.beginGroup(group_for(key));
    settings.setValue(QStringLiteral("floating"), true);
    settings.setValue(QStringLiteral("visible"), true);
    settings.setValue(QStringLiteral("geometry"),
                      QStringLiteral("%1,%2,%3,%4")
                          .arg(geometry.x())
                          .arg(geometry.y())
                          .arg(geometry.width())
                          .arg(geometry.height()));
    settings.endGroup();
    settings.sync();
}

void LayoutPersistence::save_dock(const std::string& key,
                                  const std::vector<int>& sizes) {
    QSettings& settings = bind();
    settings.beginGroup(group_for(key));
    settings.setValue(QStringLiteral("floating"), false);
    settings.setValue(QStringLiteral("visible"), true);
    settings.setValue(QStringLiteral("docked_sizes"), encode_sizes(sizes));
    settings.endGroup();
    settings.sync();
}

void LayoutPersistence::save_docked_sizes(const std::string& key,
                                          const std::vector<int>& sizes) {
    QSettings& settings = bind();
    settings.setValue(group_for(key) + QStringLiteral("/docked_sizes"),
                      encode_sizes(sizes));
    settings.sync();
}

void LayoutPersistence::save_visibility(const std::string& key,
                                        bool visible) {
    QSettings& settings = bind();
    settings.setValue(group_for(key) + QStringLiteral("/visible"), visible);
    settings.sync();
}

void LayoutPersistence::clear(const std::string& key) {
    QSettings& settings = bind();
    settings.remove(group_for(key));
    settings.sync();
}

PanelLayoutRecord LayoutPersistence::load(const std::string& key) const {
    // QSettings::beginGroup/endGroup are non-const even for reads — bind
    // non-const (the store is logically mutable state anyway).
    QSettings& settings =
        const_cast<LayoutPersistence*>(this)->bind();
    settings.beginGroup(group_for(key));
    PanelLayoutRecord record;
    record.floating = settings.value(QStringLiteral("floating"), false).toBool();
    record.visible = settings.value(QStringLiteral("visible"), true).toBool();
    record.geometry =
        parse_geometry(settings.value(QStringLiteral("geometry")).toString());
    record.docked_sizes =
        parse_sizes(settings.value(QStringLiteral("docked_sizes")).toString());
    settings.endGroup();
    return record;
}

QSettings& LayoutPersistence::bind() {
    if (settings_ != nullptr) {
        return *settings_;
    }
    if (owned_ == nullptr) {
        // One-shot legacy migration before the default backend binds — an
        // isolated LayoutPersistence (no workstation shell) still reads the
        // old panel layout.
        migrate_legacy_layout_settings();
        owned_ = std::make_unique<QSettings>(
            QString::fromLatin1(kSettingsOrg),
            QString::fromLatin1(kSettingsApp));
    }
    return *owned_;
}

const QSettings& LayoutPersistence::bind() const {
    return const_cast<LayoutPersistence*>(this)->bind();
}

QString LayoutPersistence::group_for(const std::string& key) {
    return QStringLiteral("panel_layout/") + qstr(key);
}

QString LayoutPersistence::encode_sizes(const std::vector<int>& sizes) {
    QStringList parts;
    parts.reserve(static_cast<qsizetype>(sizes.size()));
    for (const int size : sizes) {
        parts.push_back(QString::number(size));
    }
    return parts.join(QLatin1Char(','));
}

std::optional<QRect> LayoutPersistence::parse_geometry(const QString& raw) {
    const QStringList parts = raw.split(QLatin1Char(','), Qt::SkipEmptyParts);
    if (parts.size() != 4) {
        return std::nullopt;
    }
    bool ok = false;
    const int x = parts[0].toInt(&ok);
    if (!ok) return std::nullopt;
    const int y = parts[1].toInt(&ok);
    if (!ok) return std::nullopt;
    const int w = parts[2].toInt(&ok);
    if (!ok) return std::nullopt;
    const int h = parts[3].toInt(&ok);
    if (!ok) return std::nullopt;
    return QRect(x, y, w, h);
}

std::optional<std::vector<int>> LayoutPersistence::parse_sizes(
    const QString& raw) {
    const QStringList parts = raw.split(QLatin1Char(','), Qt::SkipEmptyParts);
    if (parts.isEmpty()) {
        return std::nullopt;
    }
    std::vector<int> out;
    out.reserve(parts.size());
    for (const QString& part : parts) {
        bool ok = false;
        const int value = part.toInt(&ok);
        if (!ok) {
            return std::nullopt;
        }
        out.push_back(value);
    }
    return out;
}

}  // namespace pwb::ui_shell
